File: solutions/java/0329.java
class Solution {
  public int longestIncreasingPath(int[][] matrix) {
    if (matrix.length == 0)
      return 0;

    int ans = 0;
    int[][] cached = new int[matrix.length][matrix[0].length];

    for (int i = 0; i < matrix.length; ++i)
      for (int j = 0; j < matrix[0].length; ++j)
        ans = Math.max(ans, dfs(matrix, i, j, Integer.MIN_VALUE, cached));

    return ans;
  }

  private int dfs(int[][] matrix, int i, int j, int max, int[][] cached) {
    if (i < 0 || i >= matrix.length || j < 0 || j >= matrix[0].length || matrix[i][j] <= max)
      return 0;
    if (cached[i][j] != 0)
      return cached[i][j];

    int a = dfs(matrix, i + 1, j, matrix[i][j], cached);
    int b = dfs(matrix, i - 1, j, matrix[i][j], cached);
    int c = dfs(matrix, i, j + 1, matrix[i][j], cached);
    int d = dfs(matrix, i, j - 1, matrix[i][j], cached);
    cached[i][j] = 1 + Math.max(Math.max(a, b), Math.max(c, d));

    return cached[i][j];
  }
}

File: solutions/cpp/0329.cpp
class Solution {
 public:
  int longestIncreasingPath(vector<vector<int>>& matrix) {
    if (matrix.empty()) return 0;

    const int m = matrix.size();
    const int n = matrix[0].size();

    int ans = 0;
    vector<vector<int>> memo(m, vector<int>(n));

    function<int(int, int, int)> dfs = [&](int i, int j, int maxi) {
      if (i < 0 || i == m || j < 0 || j == n) return 0;
      if (matrix[i][j] <= maxi) return 0;
      if (memo[i][j]) return memo[i][j];

      return memo[i][j] = 1 + max({dfs(i + 1, j, matrix[i][j]),
                                   dfs(i - 1, j, matrix[i][j]),
                                   dfs(i, j + 1, matrix[i][j]),
                                   dfs(i, j - 1, matrix[i][j])});
    };

    for (int i = 0; i < m; ++i)
      for (int j = 0; j < n; ++j)
        ans = max(ans, dfs(i, j, INT_MIN));

    return ans;
  }
};